resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -542,4 +728,114 @@ private void addTrace(Throwable cause) {
+     */
+    private static final class DnsServerNames {
+
+        private final String qName;
+
+        // We not expect the linked-list to be very long so a double-linked-list is overkill.
+        private ServerName head;
+        private int count;
+
+        DnsServerNames(String qName) {
+            this.qName = qName.toLowerCase(Locale.US);
+        }
+
+        void add(DnsRecord r) {
+            if (r.type() != DnsRecordType.NS || !(r instanceof DnsRawRecord)) {
+                return;
+            }
+
+            // Only include servers that serve the correct domain.
+            if (qName.length() <  r.name().length()) {
+                return;
+            }
+
+            String rName = r.name().toLowerCase(Locale.US);
+
+            int dots = 0;
+            for (int a = rName.length() - 1, b = qName.length() - 1; a >= 0; a--, b--) {
+                char c = rName.charAt(a);
+                if (qName.charAt(b) != c) {
+                    return;
+                }
+                if (c == '.') {
+                    dots++;
+                }
+            }
+
+            if (head != null && head.dots > dots) {
+                // We already have a closer match so ignore this one, no need to parse the dmainName etc.
+                return;
+            }
+
+            final ByteBuf recordContent = ((ByteBufHolder) r).content();
+            final String domainName = decodeDomainName(recordContent);
+            if (domainName == null) {
+                // Could not be parsed, ignore.
+                return;
+            }
+
+            // We are only interested in preserving the nameservers which are the closest to our qName, so ensure
+            // we drop servers that have a smaller dots count.
+            if (head == null || head.dots < dots) {
+                count = 1;
+                head = new ServerName(dots, rName, domainName);
+            } else {
+                if (head.dots == dots) {
+                    ServerName serverName = head;
+                    while (serverName.next != null) {
+                        serverName = serverName.next;
+                    }
+                    serverName.next = new ServerName(dots, rName, domainName);
+                    count++;
+                }
+            }
+        }
+
+        // Just walk the linked-list and mark the entry as removed when matched, so next lookup will need to process
+        // one node less.
+        ServerName remove(String name) {
+            ServerName serverName = head;
+
+            while (serverName != null) {
+                if (!serverName.removed && serverName.name.equalsIgnoreCase(name)) {
+                    serverName.removed = true;
+                    return serverName;
+                }
+                serverName = serverName.next;
+            }
+            return null;
+        }
+
+        int size() {
+            return count;
+        }
+
+        static final class ServerName {
```

Comment:
We have `qName` (questionName?), `rName` (recordName?), and `sName` (serverName?) all being used. Is it possible to name this class something more descriptive to clarify its usage?